File: StopWatch_oc/StopWatch_oc/ViewController.m
//
//  ViewController.m
//  StopWatch
//
//  Created by ynfMac on 2019/5/6.
//  Copyright © 2019年 ynfMac. All rights reserved.
//

#import "ViewController.h"
#import "WSButton.h"
#import "UIImage+color.h"
#import "StopWatchData.h"
#import "recordTimeCell.h"

@interface ViewController ()<UITableViewDelegate,UITableViewDataSource>
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UILabel *timeLabel;
@property (weak, nonatomic) IBOutlet UIButton *resetButton;
@property (weak, nonatomic) IBOutlet UIButton *startButton;

@property(strong,nonatomic,readonly)recordTimeCell * currentRunTimeCell;

@end

static NSString *const cellId = @"cellId";

@implementation ViewController
{
    NSTimer *_timer;
    NSTimeInterval _totalTime;
    NSTimeInterval _interval;
    StopWatchData *_timeData;
    
    recordTimeCell * _currentRunTimeCell; // 正在计时的cell
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self _initVariables];
    
    [self _setupNavigation];
    
    [self _setupButtons];
    
}

- (void)_initVariables{
    
    _totalTime = 0;
    _interval = 0;
    
    _timeLabel.text = @"00:00.00";
    
    _timeData = [[StopWatchData alloc] init];
}

- (void)_setupNavigation{
    
    // 修改导航栏标题的颜色
    [self.navigationController.navigationBar setTitleTextAttributes:@{NSForegroundColorAttributeName:[UIColor whiteColor],
                                                                                 NSFontAttributeName:[UIFont systemFontOfSize:18]
                                                                      }];
    
    // 修改导航栏的背景颜色
    [self.navigationController.navigationBar setBarTintColor:[UIColor blackColor]];
}

- (void)_setupButtons{
    
    // resetButton
    self.resetButton.layer.cornerRadius = self.resetButton.bounds.size.width * 0.5;
    self.resetButton.layer.masksToBounds = YES;
    
    [self.resetButton setTitle:@"计次" forState:UIControlStateDisabled];
    self.resetButton.enabled = NO;
    [self.resetButton setBackgroundImage:[UIImage ws_imageWithColor:[UIColor colorWithRed:0.08 green:0.08 blue:0.08 alpha:1]] forState:UIControlStateDisabled];
    
    [self.resetButton setTitle:@"复位" forState:UIControlStateSelected];
    [self.resetButton setTitleColor:[UIColor whiteColor] forState:UIControlStateSelected];
    [self.resetButton setBackgroundImage:[UIImage ws_imageWithColor:[UIColor grayColor]] forState:UIControlStateSelected];
    
    [self.resetButton setTitle:@"计次" forState:UIControlStateNormal];
    [self.resetButton setBackgroundImage:[self.resetButton backgroundImageForState:UIControlStateSelected] forState:UIControlStateNormal];
    
    // startButton
    self.startButton.layer.cornerRadius = self.resetButton.bounds.size.width * 0.5;
    self.startButton.layer.masksToBounds = YES;
    
    [self.startButton setBackgroundImage:[UIImage ws_imageWithColor:[UIColor colorWithRed:20 / 255.0 green:40 / 255.0 blue:24 / 255.0 alpha:1]] forState:UIControlStateNormal];
    [self.startButton setTitle:@"启动" forState:UIControlStateNormal];
    [self.startButton setTitleColor:[UIColor colorWithRed:64 / 255.0 green:203 / 255.0 blue:96 / 255.0 alpha:1] forState:UIControlStateNormal];
    
    [self.startButton setBackgroundImage:[UIImage ws_imageWithColor:[UIColor colorWithRed:74 / 255.0 green:16 / 255.0 blue:17 / 255.0 alpha:1]] forState:UIControlStateSelected];
    [self.startButton setTitle:@"停止" forState:UIControlStateSelected];
    [self.startButton setTitleColor:[UIColor redColor] forState:UIControlStateSelected];

}

- (IBAction)reset:(WSButton *)sender {
    if (sender.selected) { // 复位
        
        //数据全部清零
        [self _initVariables];
        
        // 刷新tableview
        [_tableView reloadData];
        
        // 切换按钮状态
        sender.enabled = NO;
        sender.selected = NO;
        
    } else { // 计次
        
        _interval = 0;
        [_timeData beginingNewTimeWithCompletion:nil];
    }
    
    [_tableView reloadData];
    
}

- (IBAction)start:(WSButton *)sender {
    sender.selected = !sender.selected;
    
    if (sender.selected) {
        
        if (!self.resetButton.enabled){
            self.resetButton.enabled = YES;
        } else {
            self.resetButton.selected = NO;
        }
        
        if (_timeData.isReset) {
            [_timeData beginingNewTimeWithCompletion:^{
                [self->_tableView reloadData];
            }];
        }
        
        _timer = [NSTimer scheduledTimerWithTimeInterval:0.01 repeats:YES block:^(NSTimer * _Nonnull timer) {
            self->_totalTime++;
            self->_interval++;
            self->_timeLabel.text = [self _convertTime:self->_totalTime];
            
            if (self->_currentRunTimeCell != [self _topCell]) {
                self->_currentRunTimeCell = [self _topCell];
            }
            
            self->_currentRunTimeCell.detailTextLabel.text = [self _convertTime:self->_interval];
            [self->_timeData timing:self->_interval];
            
        }];
        [[NSRunLoop currentRunLoop] addTimer:_timer forMode:NSRunLoopCommonModes];
    } else {
        self.resetButton.selected = YES;
        [_timer invalidate];
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    
    recordTimeCell *cell = [tableView dequeueReusableCellWithIdentifier:cellId];
    if (!cell) {
        cell = [[recordTimeCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:cellId];
    }
    
    [cell recordTimeWithData:_timeData atIndexPath:indexPath];

    return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return _timeData.times.count;
}

- (recordTimeCell *)_topCell{
    return [_tableView cellForRowAtIndexPath:[NSIndexPath indexPathForRow:0 inSection:0]];
}

- (NSString *)_convertTime:(NSUInteger)time{
    return [NSString stringWithFormat:@"%02ld:%02ld.%02ld",time / 100 / 60, time / 100 % 60, time % 100];;
}


@end

File: stopWatch_swift/stopWatch_swift/ViewController.swift
//
//  ViewController.swift
//  stopWatch_swift
//
//  Created by ynfMac on 2019/7/8.
//  Copyright © 2019 ynfMac. All rights reserved.
//

import UIKit

class ViewController: UIViewController,UITableViewDelegate,UITableViewDataSource{

    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var timeLable: UILabel!
    @IBOutlet weak var resetButton: WSButton!
    @IBOutlet weak var startButton: WSButton!
    
    var totalTime = 0 {
        willSet{
            timeLable.text = convertTime(seconds: newValue)
        }
    }
    var intervals = 0 {
        willSet{
            let cell = timingCell()
            cell?.detailTextLabel?.text = convertTime(seconds: newValue)
        }

    }
    
    var timer: Timer?
    let data = stopWatchData()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setNavigationBar()
        
        setUI()
    }
    
    @IBAction func reset(_ sender: WSButton) {
        if sender.isSelected {
            
            resetVariables()
            
            sender.isEnabled = false
            sender.isSelected = false
        } else {
            intervals = 0
            data.beginingNewTime()
        }
        
        tableView.reloadData()
    }
    
    func resetVariables(){
        totalTime = 0
        intervals = 0
        data.reset()
    }
    
    @IBAction func start(_ sender: WSButton) {
        sender.isSelected = !sender.isSelected
        
        if !sender.isSelected {
            resetButton.isSelected = true
            timer?.invalidate()
            timer = nil
            return
        }
        
        if data.isReset {
            data.beginingNewTime {
                self.tableView.reloadData()
                self.resetButton.isEnabled = true
            }
        } else {
            resetButton.isSelected = false
        }
        
        if timer == nil {
            timer = Timer.init(timeInterval: 0.01, repeats: true, block: { [weak self] _ in
                self?.totalTime += 1
                self?.intervals += 1
                self?.data.timing(time: self!.intervals)
            })
            RunLoop.current.add(self.timer!, forMode: .commonModes)
        }
    }
    
    func setNavigationBar(){
        navigationController?.navigationBar.titleTextAttributes = [.foregroundColor: UIColor.white,
                                                                   .font:UIFont.systemFont(ofSize: 18)]
        navigationController?.navigationBar.barTintColor = UIColor.black
    }
    
    func setUI(){
        resetButton.layer.cornerRadius = resetButton.bounds.size.width * 0.5
        resetButton.layer.masksToBounds = true
        
        resetButton.setTitle("计次", for: .disabled)
        resetButton.isEnabled = false
        resetButton .setBackgroundImage(createImage(UIColor.init(red: 0.08, green: 0.08, blue: 0.08, alpha: 1)), for: .disabled)
        
        resetButton.setTitle("复位", for: .selected);
        resetButton.setTitleColor(UIColor.white, for: .selected)
        resetButton.setBackgroundImage(createImage(UIColor.gray), for: .selected)
        
        resetButton.setTitle("计次", for: .normal)
        resetButton.setBackgroundImage(resetButton.backgroundImage(for: .selected), for: .normal)
        
        // startButton
        startButton.layer.cornerRadius = self.resetButton.bounds.size.width * 0.5;
        startButton.layer.masksToBounds = true;
        
        startButton.setBackgroundImage(createImage(UIColor.init(red: 20 / 255.0, green: 40 / 255.0, blue: 24 / 255.0, alpha: 1)), for: .normal)
    
        startButton.setTitle("启动", for: .normal)
        startButton.setTitleColor(UIColor.init(red: 64 / 255.0, green: 203 / 255.0, blue: 96 / 255.0, alpha: 1), for: .normal)
        
        startButton.setBackgroundImage(createImage(UIColor.init(red: 74 / 255.0, green: 16 / 255.0, blue: 17 / 255.0, alpha: 1)), for: .selected)
        startButton.setTitle("停止", for: .selected)
        startButton.setTitleColor(.red, for: .selected);
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle{
        return .lightContent
    }
    
    func createImage(_ color: UIColor) -> UIImage{
        let rect = CGRect.init(x: 0.0, y: 0.0, width: 1.0, height: 1.0)
        UIGraphicsBeginImageContext(rect.size)
        let context = UIGraphicsGetCurrentContext()
        context?.setFillColor(color.cgColor)
        context?.fill(rect)
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return image!
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return data.times.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        
        var cell = tableView.dequeueReusableCell(withIdentifier: "cell") as? RecordTimeCell
        
        if cell == nil {
            cell = RecordTimeCell.init(style: .value1, reuseIdentifier: "cell")
        }
        
        cell?.recordTime(with: data, indexPath)
         
        return cell!;
    }
    
    func timingCell()->RecordTimeCell?{
        return tableView.cellForRow(at: NSIndexPath(row: 0, section: 0) as IndexPath) as? RecordTimeCell
    }
    
    func convertTime(seconds: Int) ->String{
        return String(format: "%02ld:%02ld.%02ld",seconds / 100 / 60, seconds / 100 % 60, seconds % 100)
    }
}
